aidb/engine/base_engine.py

```diff
@@ -358,3 +358,110 @@ def _get_left_join_str(self, rep_table_name: str, tables: List[str]) -> str:
       else:
         raise Exception(f'Can\'t join table {rep_table_name} and {table_name}')
     return f'FROM {rep_table_name}\n' + '\n'.join(join_strs)
+
+
+  def _call_user_function(self, res_df: pd.DataFrame, function_name: str, args_list: List[str]):
+    function_name = str.lower(function_name)
+    if function_name not in self._config.user_defined_function:
+      raise Exception(f'{function_name} has not been added into config, please add it first.')
+    parameter_list = []
+    for args in args_list:
+      parameter_list.append(res_df[args])
+
+    function_results = self._config.user_defined_function[function_name](*parameter_list)
+    # for result that has dataframe format, change it to a list of tuples
+    if isinstance(function_results, pd.DataFrame):
+      res_list_of_tuple = [tuple(row) for row in function_results.itertuples(index=False)]
```

Comment:
Why do you need the tuple cast? itertuples already returns a tuple I thought

Reply:
Parameter 'row' is of the type 'pandas.core.frame.Pandas', so we need to cast it to the tuple type.

Reply:
hmm... is there a way to get a native tuple type? I wonder if we should consider polars... probably not worth switching though